### PR TITLE
Feat/add steep types

### PR DESCRIPTION
**Pull request by aseure**

### CODE REVIEW

**Review by chloelbn — APPROVED** [review]

File: lib/algolia/transport/transport.rb [hunk 1]
```diff
@@ -127,15 +127,12 @@ def build_body(body, request_options, method)
...
       #
       # @return [Hash] merged headers
       #
-      def generate_headers(request_options = {})
-        headers                                                     = {}
-        extra_headers                                               = request_options.headers || {}
-        @config.default_headers.each { |key, val| headers[key.to_s] = val }
-        extra_headers.each { |key, val| headers[key.to_s]           = val }
+      def generate_headers(request_options)
+        headers = @config.default_headers.merge(request_options.headers)
```
Comment (DevinCodes) [inline lib/algolia/transport/transport.rb:135]:
```suggestion
        headers = @config.default_headers.merge(request_options.headers || {})
```

File: sig/responses/add_api_key_response.rbs [hunk 2]
```diff
@@ -0,0 +1,9 @@
+module Algolia
+  class AddApiKeyResponse < BaseResponse
+    attr_reader raw_response: Hash[Symbol, String|[String]]
+
+    # TODO add client
```
Comment (DevinCodes) [inline sig/responses/add_api_key_response.rbs:5]: I think we can address all the `# TODO`s in this PR, right? I believe we'll have (almost) all types ready now.

File: sig/search_client.rbs [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+      def self.create_with_config: (Search::Config config) -> self
+      def wait_task: (String index_name, Integer task_id, ?Integer time_before_retry, ?::Hash[Symbol, String|[String]] opts) -> nil
+      def get_task_status: (String index_name, Integer task_id, ?::Hash[Symbol, String|[String]] opts) -> String
+
+      # TODO add Index
+      #
+      def init_index: (String index_name) -> untyped
+      def list_indexes: (?::Hash[Symbol, String|[String]] opts) -> Hash[Symbol, String|[String]]
+      def get_logs: (?::Hash[Symbol, String|[String]] opts) -> Hash[Symbol, String|[String]]
+      def copy_rules: (String src_index_name, String dest_index_name, ?::Hash[Symbol, String|[String]] opts) -> IndexingResponse
```
Comment (DevinCodes) [inline sig/search_client.rbs:19]: IIRC, we wanted to create a type for this `?::Hash[Symbol, String|[String]] opts`. Did you find anything for this? 🙂

File: sig/search_client.rbs [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+      def get_user_id: (String user_id, ?::Hash[Symbol, String|[String]] opts) -> Hash[Symbol, String|[String]]
+      def list_clusters: (?::Hash[Symbol, String|[String]] opts) -> Hash[Symbol, String|[String]]
+      def list_user_ids: (?::Hash[Symbol, String|[String]] opts) -> Hash[Symbol, String|[String]]
+      def remove_user_id: (String user_id, ?::Hash[Symbol, String|[String]] opts) -> Hash[Symbol, String|[String]]
+      def search_user_ids: (String query, ?::Hash[Symbol, String|[String]] opts) -> Hash[Symbol, String|[String]]
+      def pending_mappings?: (?::Hash[Symbol, String|[String]] opts) -> Hash[Symbol, String|[String]]
+      def custom_request: (Hash[Symbol, String|[String]] data, String uri, Symbol method, Integer call_type, ?::Hash[Symbol, String|[String]] opts) -> Hash[Symbol, String|[String]]
+      end
+      end
+      end
```
Comment (DevinCodes) [inline sig/search_client.rbs:57]:
```suggestion
   end
end
```

**Review by DevinCodes — CHANGES_REQUESTED** [review]

File: sig/responses/add_api_key_response.rbs (hunk 2 above)
Comment (chloelbn) [reply sig/responses/add_api_key_response.rbs:5]: I'll address them in a different PR! This only mainly serves as putting back the lost commits 😢

File: sig/search_client.rbs (hunk 3 above)
Comment (chloelbn) [reply sig/search_client.rbs:19]: not yet! will do it later once the branch is clean again

**Review by DevinCodes — CHANGES_REQUESTED** [review]

**Comment by aseure** [comment]
Closing this one so that I don't have it listed in my personal opened PRs anymore now that I don't work for Algolia anymore.

**Pull request closed without merging**